CONTRIBUTING_CODE.md

```diff
@@ -0,0 +1,284 @@
+
+Here are the key sections in this docs:
+
+- [Getting Set up](#getting-set-up)
+- [What is in the TypeScript repo?](#what-is-in-the-typescript-repo)
+- [The TypeScript Compiler](#the-typescript-compiler)
+- [The TSServer](#the-tsserver)
+- [Tips and Tricks](#tips-and-tricks)
+
+## Getting set up
+
+You need a modern version [of Node.js](https://nodejs.org/en/) installed and then:
+
+- Clone the repo: `git clone https://github.com/microsoft/TypeScript` (optional: `--depth=1`)
+- `cd TypeScript`
+- `npm i` - installs dependencies
+- `npm run gulp` - builds the TypeScript compiler
+- `code .` then briefly come back to terminal to run:
+- `npm test` which will verify everything works fine
+
+With that done, you should have a copy of the compiler up and running with tests running in the background to
+prove that everything is working correctly. This could take around 10m, you should see be a progress indicator in
+your terminal.
+
+As an alternative, this repository also includes a [development container](https://code.visualstudio.com/docs/remote/containers) that you can use to quickly create an isolated development environment with all the tools you need to start working on TypeScript. To get started with a dev container and VS Code, either:
+
+- Clone the TypeScript repository locally and use the `Open Folder in Container` command.
+- Use the `Clone Repository in Container Volume` command to clone the TypeScript repository into a new container.
+
+### Build commands
+
+You can use `npm run gulp -- --tasks --depth 1` to see all the possible build commands.
+
+### What happened during setup
+
+TypeScript is built in TypeScript. The compiler uses a "last known good" version of the TypeScript to bootstrap,
+you can find this version in `/lib`. The command `npm run gulp` created a copy of the compiler at `built/local`.
+The corresponding version of the `tsc` compiler used in the npm package is now available via:
+`node built/local/tsc.js --help`.
+
+TypeScript has a _comprehensive_ test suite, most of the tests are high level integration tests which you can find
+at `tests/cases`. Given the scope and depth of a compiler like TypeScript, we use a lot of tests to ensure
+improvements don't break old requirements.
+
+### Configuring your editor
+
+To get your VS Code environment working smoothly, set up the per-user config:
+
+- Set up your `.vscode/launch.json` by running: `cp .vscode/launch.template.json .vscode/launch.json`
+- Set up your `.vscode/settings.json` by running: `cp .vscode/settings.template.json .vscode/settings.json`
+
+
+## What is in the TypeScript repo?
+
+The TypeScript repo is a single codebase which creates a set of files which are released as the "typescript" npm
+module:
+
+- `typescript.js` - The compiler API
+- `tsc.js` - The command line experience for using TypeScript, which uses `typescript.js`
+- `typescriptServices.js` - A web server which responds to questions about TS/JS code from editors, which uses
```

Comment:
Curious - who uses this, and what do they use it for?

Reply:
Yes

Reply:
It is just the services layer, in fact, it's nearly identical to `typescript.js`, the only difference being it has a UMD wrapper for browser embedding (and exposes a global `ts` in such an environment).